feat: support header parameters

DDtKey opened the pull request:
Closes #255
Base support for Header parameters.

And small additional change: use `Vec::with_capacity` instead of `Vec::new`

All current tests passes, but it would be nice to add some test with `header` params (I tested it on my own examples). 
I've not found `Header` in `dropshot` to write "specific-test". So we can choose any public openapi spec to add. (Upd: see https://github.com/oxidecomputer/progenitor/pull/210#issuecomment-1298794146)

DDtKey commented:
@ahl could you take a look?  (sorry for bothering, i don't know who is maintainer)

ahl commented:
Thanks for this. I'm going to take a look, but I wanted to set the expectation that it might be a week or two as I'm a little slammed at the moment.

jayvdb commented:
IMO it would be helpful to add a header to the files in https://github.com/oxidecomputer/progenitor/tree/main/sample_openapi , esp `keeper.json`, and update the https://github.com/oxidecomputer/progenitor/tree/main/example-build if required to make this work.

DDtKey commented:
@jayvdb I totally agree, but I didn't want to change existed APIs because I'm not sure if they really has any headers as params in their original specs. 

I've checked and found that `keeper` has required `Authorization` header ([link](https://github.com/jclulow/keeper/blob/b32a1f39ed1c54123662cf405e8e0ab3299fc8d2/server/src/main.rs#L69)), so I've added it to samples and updated tests.

jayvdb commented:
fyi, for non-mandatory headers, I see

```rust
rror[E0277]: `?` couldn't convert the error to `progenitor_client::Error`
   --> clients/foo/src/lib.rs:501:78
    |
501 |             header_map.append("X-Foo-ApiVersion", HeaderValue::try_from(v)?);
    |                                                                              ^ the trait `std::convert::From<reqwest::header::InvalidHeaderValue>` is not implemented for `progenitor_client::Error`
    |
    = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = help: the following other types implement trait `std::convert::From<T>`:
              <progenitor_client::Error<E> as std::convert::From<reqwest::Error>>
    = note: required for `std::result::Result<progenitor_client::ResponseValue<types::ApplicationRegistrationInfo>, progenitor_client::Error>` to implement `std::ops::FromResidual<std::result::Result<std::convert::Infallible, reqwest::header::InvalidHeaderValue>>`
```

jayvdb reviewed (approved):
WFM. Thanks.

DDtKey commented:
Gentle ping @ahl

DDtKey commented:
Gentle ping @ahl

ahl reviewed (commented):
looks good; apologies for the long delay

ahl commented on progenitor-impl/src/method.rs:
```diff
@@ -734,6 +765,45 @@ impl Generator {
+            .filter_map(|param| match &param.kind {
+                OperationParameterKind::Header(required) => {
+                    let qn = &param.api_name;
+                    let qn_ident = format_ident!("{}", &param.name);
```
I think this stood for "query name", but this is fine...

DDtKey replied:
Oh, yeah. Changed to `hn` to keep the naming style

ahl reviewed (changes requested):
can you please add a note to the CHANGELOG.md file?

DDtKey commented:
> can you please add a note to the CHANGELOG.md file?

Well, I see only [this file](https://github.com/oxidecomputer/progenitor/blob/main/CHANGELOG.adoc) and it tells that it will be updated automatically on release. Isn't? 🤔

ahl commented:
> > can you please add a note to the CHANGELOG.md file?
> 
> Well, I see only [this file](https://github.com/oxidecomputer/progenitor/blob/main/CHANGELOG.adoc) and it tells that it will be updated automatically on release. Isn't? 🤔

The sections are updated; the content is added by hand; check out the history for the file.

DDtKey commented:
> > > can you please add a note to the CHANGELOG.md file?
> 
> > 
> 
> > Well, I see only [this file](https://github.com/oxidecomputer/progenitor/blob/main/CHANGELOG.adoc) and it tells that it will be updated automatically on release. Isn't? 🤔
> 
> 
> 
> The sections are updated; the content is added by hand; check out the history for the file.

Oh, got it. Check please.
A bit confusing to be honest, probably only to me.

ahl reviewed (approved):
nice

ahl commented:
`EXPECTORATE=overwrite cargo test`

DDtKey commented:
> `EXPECTORATE=overwrite cargo test`

Yeah, sorry for some mess here. I'm just not near to my main laptop. So I don't have all of tools. 

It should be fine now, just pls use squashing merge strategy if CI is green again 😅

DDtKey commented:
@ahl updated, all checks are passed locally (was finally able to return home)

Sorry for the mess one more time

DDtKey commented:
@ahl could you approve CI runs one more time? I believe we can merge this PR after last changes

ahl commented:
rerunning a failed macos CI run that appeared to be a transient issue

ahl commented:
woot! Thanks for taking the long slow walk to get this done @DDtKey!
